std/experimental/color/rgb.d

```diff
@@ -0,0 +1,852 @@
+    BruceRGB,
+    CIE_RGB,
+    ColorMatchRGB,
+    DonRGB4,
+//    ECI_RGB_v2,
+    EktaSpacePS5,
+    NTSC_RGB,
+    PAL_SECAM_RGB,
+    ProPhotoRGB,
+    SMPTE_C_RGB,
+    WideGamutRGB
++/
+}
+
+
+struct RGB(string components_, ComponentType_, bool linear_ = false, RGBColorSpace colorSpace_ = RGBColorSpace.sRGB) if(isValidComponentType!ComponentType_)
+{
+@safe: pure: nothrow: @nogc:
+
+    // RGB colors may only contain components 'rgb', or 'l' (luminance)
+    // They may also optionally contain an 'a' (alpha) component, and 'x' (unused) components
+    static assert(AllIn!("rgblax", components), "Invalid Color component '"d ~ NotIn!("rgblax", components) ~ "'. RGB colors may only contain components: r, g, b, l, a, x"d);
+    static assert(AnyIn!("rgbal", components), "RGB colors must contain at least one component of r, g, b, l, a.");
+    static assert(!CanFind!(components, 'l') || !AnyIn!("rgb", components), "RGB colors may not contain rgb AND luminance components together.");
+
+    // create members for some useful information
+    alias ComponentType = ComponentType_;
+    enum string components = components_;
+    enum RGBColorSpace colorSpace = colorSpace_;
+    enum bool linear = linear_;
+    enum bool hasAlpha = CanFind!(components, 'a');
+
+
+    // mixin will emit members for components
+    template Components(string components)
+    {
+        static if(components.length == 0)
+            enum Components = "";
+        else
+            enum Components = (components[0] == 'x' ? "private " : "") ~ ComponentType.stringof ~ ' ' ~ components[0] ~ " = 0;\n" ~ Components!(components[1..$]);
+    }
+    mixin(Components!components);
+
+    // missing components will be fabricated such that code that references them compiles
+    mixin template MissingComponent(char name)
+    {
+        // missing color components will equal the luminance, if it exists. 0 otherwise.
+        static if(!CanFind!(components, 'l'))
+            mixin("enum ComponentType "~name~" = 0;");
+        else
+            mixin("alias "~name~" = l;");
+    }
+    static if(!CanFind!(components, 'r'))
+        mixin MissingComponent!'r';
+    static if(!CanFind!(components, 'g'))
+        mixin MissingComponent!'g';
+    static if(!CanFind!(components, 'b'))
+        mixin MissingComponent!'b';
+    static if(!CanFind!(components, 'l'))
+        @property ComponentType l() const { return toGrayscale!colorSpace(r, g, b); }
+    static if(!CanFind!(components, 'a'))
+        enum ComponentType a = 0;
+
+    // functions that return the color channels as a tuple
+    static if(CanFind!(components, 'l'))
+        @property auto tristimulus() const { auto t = l; return tuple(t, t, t); }
+    else
+        @property auto tristimulus() const { return tuple(r, g, b); }
+    @property auto tristimulusWithAlpha() const { return tuple(tristimulus.expand, a); }
+
+    // RGB/A initialiser
+    this(ComponentType r, ComponentType g, ComponentType b, ComponentType a = 0)
```

Comment:
Alpha isn't a synonym for opacity, that's just a typical application. I've also used it for countless other things.
It just feels weird that by default, you have to be careful that an argument you *didn't* write may damage your alpha values when performing typical math.
Usually you're probably only acting on the rgb channels, you shouldn't expect the alpha channel would be changed unless you wrote some value to change it.